txnBuckets.go

```diff
@@ -0,0 +1,121 @@
+package turtleDB
+
+import "sync"
+
+func newTxnBuckets(b *buckets) *txnBuckets {
+	var tb txnBuckets
+	tb.b = b
+	tb.m = make(map[string]*txnBucket)
+	return &tb
+}
+
+type txnBuckets struct {
+	mux sync.RWMutex
+
+	// Main branch
+	b *buckets
+	// Transaction buckets
+	m map[string]*txnBucket
+}
+
+// create will create a bucket at a given key
+func (tb *txnBuckets) create(key string, rb *bucket) (bkt *txnBucket, created bool) {
+	var ok bool
+	if bkt, ok = tb.m[key]; !ok {
+		// Get reference bucket
+		if rb == nil {
+			rb, _ = tb.b.get(key)
+		}
+
+		bkt = newTxnBucket(rb)
+		tb.m[key] = bkt
+		created = true
+	}
+
+	return
+}
+
+// Get will get a bucket
+func (tb *txnBuckets) Get(key string) (b Bucket, err error) {
+	var ok bool
+	tb.mux.RLock()
+	b, ok = tb.m[key]
+	tb.mux.RUnlock()
+
+	if ok {
+		return
+	}
+
+	tb.mux.Lock()
+	// Check again to ensure the txnBucket wasn't created inbetween locks
+	if b, ok = tb.m[key]; !ok {
+		if bb, err := tb.b.get(key); err == nil {
+			tbkt := newTxnBucket(bb)
+			tb.m[key] = tbkt
+			b = tbkt
+			ok = true
+		}
+	}
+	tb.mux.Unlock()
+
+	if !ok {
+		// No match was found, return error
+		err = ErrKeyDoesNotExist
+	}
+
+	return
+}
+
+// Create will create and return a bucket
+func (tb *txnBuckets) Create(key string) (bkt Bucket, err error) {
+	tb.mux.Lock()
+	defer tb.mux.Unlock()
+	bkt, _ = tb.create(key, nil)
+	return
+}
+
+// Delete will delete a bucket
+func (tb *txnBuckets) Delete(key string) (err error) {
+	var (
+		bkt *txnBucket
+		ok  bool
+	)
+
+	// Lock before doing anything
+	tb.mux.Lock()
+	defer tb.mux.Unlock()
+
+	if bkt, ok = tb.m[key]; !ok {
+		// No bucket exists at this key, no need to delete it
+		return ErrKeyDoesNotExist
+	}
+
+	// Delete the contents of the bucket
+	bkt.deleteAll()
+
+	// TODO: Add actual removal of the bucket, rather than just it's contents
```

Comment:
Agreed, I think I decided on the way I wanted to approach this. I was giving myself a few days to make sure I attacked it correctly. I'll have an implementation of this sometime later today.